ccl/serverccl: TestTenantCannotSetClusterSetting failed

ccl/serverccl.TestTenantCannotSetClusterSetting [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3429501&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3429501&tab=artifacts#/) on master @ [6eb9df1f3d460f904b5254c3d32c745eecbedd3f](https://github.com/cockroachdb/cockroach/commits/6eb9df1f3d460f904b5254c3d32c745eecbedd3f):

Fatal error:

```
panic: tenant-usage-select: context canceled

```
Stack: 

```
goroutine 23204 [running]:
github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostserver.(*instance).TokenBucketRequest.func1(0x69b59b0, 0xc002a9d0b0, 0xc002b80fd0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go:83 +0x60a
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn.func1(0x69b59b0, 0xc002a9d0b0, 0xc002b80fd0, 0xc00131a5a0, 0x3ff0000000000000)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:818 +0x43
github.com/cockroachdb/cockroach/pkg/kv.(*Txn).exec(0xc002b80fd0, 0x69b59b0, 0xc002a9d0b0, 0xc001da54c0, 0xc002b80fd0, 0xc00383b700)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:879 +0xdc
github.com/cockroachdb/cockroach/pkg/kv.(*DB).Txn(0xc007663000, 0x69b59b0, 0xc002a9d0b0, 0xc0059715c0, 0xc002a9d170, 0xc002a9d1a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/db.go:817 +0x105
github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostserver.(*instance).TokenBucketRequest(0xc000cbec60, 0x69b59b0, 0xc002a9d0b0, 0xa, 0xc000c2c690, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go:54 +0x2ff
github.com/cockroachdb/cockroach/pkg/server.(*Node).TokenBucket(0xc0072b9500, 0x69b59b0, 0xc002a9d0b0, 0xc000c2c690, 0xc0072b9500, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:1403 +0x18e
github.com/cockroachdb/cockroach/pkg/roachpb._Internal_TokenBucket_Handler.func1(0x69b59b0, 0xc002a9d0b0, 0x50c2000, 0xc000c2c690, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:8787 +0x89
github.com/cockroachdb/cockroach/pkg/util/tracing.ServerInterceptor.func1(0x69b59b0, 0xc002a9d0b0, 0x50c2000, 0xc000c2c690, 0xc00383b6e0, 0xc000cd0780, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/grpc_interceptor.go:126 +0x529
google.golang.org/grpc.getChainUnaryHandler.func1(0x69b59b0, 0xc002a9d0b0, 0x50c2000, 0xc000c2c690, 0xc003a66988, 0xb6c208, 0x4f7d800, 0xc003ca7280)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1134 +0xe7
github.com/cockroachdb/cockroach/pkg/rpc.NewServer.func1(0x69b59b0, 0xc002a9d0b0, 0x50c2000, 0xc000c2c690, 0xc00383b6e0, 0xc003ca7280, 0xc003ca7280, 0x6, 0x49e3460, 0xc003b9dc40)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/context.go:177 +0xab
google.golang.org/grpc.getChainUnaryHandler.func1(0x69b59b0, 0xc002a9d0b0, 0x50c2000, 0xc000c2c690, 0x50c2000, 0xc000c2c690, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1134 +0xe7
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.unaryInterceptor(0x1, 0x69b59b0, 0xc002a9d0b0, 0x50c2000, 0xc000c2c690, 0xc00383b6e0, 0xc003ca7240, 0xc003a66ac8, 0xb6c208, 0x4f7d800, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/rpc/auth.go:71 +0xc3
google.golang.org/grpc.chainUnaryServerInterceptors.func1(0x69b59b0, 0xc002a9d020, 0x50c2000, 0xc000c2c690, 0xc00383b6e0, 0xc000cd0780, 0xc003a66ba0, 0x568b86, 0x4f9cc60, 0xc002a9d020)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1120 +0xd0
github.com/cockroachdb/cockroach/pkg/roachpb._Internal_TokenBucket_Handler(0x5248a40, 0xc0072b9500, 0x69b59b0, 0xc002a9d020, 0xc0029c3440, 0xc007bbd5e0, 0x69b59b0, 0xc002a9d020, 0xc00409be00, 0x13)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/api.pb.go:8789 +0x150
google.golang.org/grpc.(*Server).processUnaryRPC(0xc007eac000, 0x6a20678, 0xc000a7c480, 0xc00131a480, 0xc0036d4750, 0x8dcde48, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1293 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc007eac000, 0x6a20678, 0xc000a7c480, 0xc00131a480, 0x0)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:1618 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc005996f70, 0xc007eac000, 0x6a20678, 0xc000a7c480, 0xc00131a480)
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:941 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/github.com/cockroachdb/cockroach/vendor/google.golang.org/grpc/server.go:939 +0x1fd
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestTenantCannotSetClusterSetting
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestTenantCannotSetClusterSetting557345286
    test_log_scope.go:74: use -show-logs to present logs inline
```
</p>
</details>
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestTenantCannotSetClusterSetting PKG=./pkg/ccl/serverccl TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/server asubiotto
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantCannotSetClusterSetting.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/multitenantccl/tenantcostserver/BUILD.bazel
+++ b/pkg/ccl/multitenantccl/tenantcostserver/BUILD.bazel
@@ -24,6 +24,7 @@ go_library(
         "//pkg/sql/sem/tree",
         "//pkg/sql/sessiondata",
         "//pkg/util",
+        "//pkg/util/log",
         "//pkg/util/metric",
         "//pkg/util/metric/aggmetric",
         "//pkg/util/syncutil",

--- a/pkg/ccl/multitenantccl/tenantcostserver/system_table.go
+++ b/pkg/ccl/multitenantccl/tenantcostserver/system_table.go
@@ -24,6 +24,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
 	"github.com/cockroachdb/cockroach/pkg/util"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/errors"
 )
 
@@ -392,16 +393,16 @@ func (h *sysTableHelper) accomodateNewInstance(tenant *tenantState, instance *in
 
 // maybeCheckInvariants checks the invariants for the system table with a random
 // probability and only if this is a test build.
-func (h *sysTableHelper) maybeCheckInvariants() error {
+func (h *sysTableHelper) maybeCheckInvariants(ctx context.Context) error {
 	if util.CrdbTestBuild && rand.Intn(10) == 0 {
-		return h.checkInvariants()
+		return h.checkInvariants(ctx)
 	}
 	return nil
 }
 
 // checkInvariants reads all rows in the system table for the given tenant and
 // checks that the state is consistent.
-func (h *sysTableHelper) checkInvariants() error {
+func (h *sysTableHelper) checkInvariants(ctx context.Context) error {
 	// Read the two rows for the per-tenant state (instance_id = 0) and the
 	// per-instance state.
 	rows, err := h.ex.QueryBufferedEx(
@@ -430,7 +431,12 @@ func (h *sysTableHelper) checkInvariants() error {
 		h.tenantID.ToUint64(),
 	)
 	if err != nil {
-		return err
+		if h.ctx.Err() == nil {
+			log.Warningf(ctx, "checkInvariants query failed: %v", err)
+		}
+		// We don't want to cause a panic for a query error (which is expected
+		// during shutdown).
+		return nil
 	}
 	if len(rows) == 0 {
 		return nil

--- a/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go
+++ b/pkg/ccl/multitenantccl/tenantcostserver/token_bucket.go
@@ -79,7 +79,7 @@ func (s *instance) TokenBucketRequest(
 			return err
 		}
 
-		if err := h.maybeCheckInvariants(); err != nil {
+		if err := h.maybeCheckInvariants(ctx); err != nil {
 			panic(err)
 		}
 		consumption = tenant.Consumption

